### PR TITLE
Move retrieve_counter_data to Phylanx

### PR DESCRIPTION
**Pull request by parsa**
This PR moves `retrieve_counter_data()` to Phylanx and adds a test for it.

### CODE REVIEW

**Comment by parsa** [comment]
I noticed `primitive_counter` test has been broken since it was missing `return hpx::util::report_errors()`. Now that I added it the test fails. Needs investigating.

**Comment by parsa** [comment]
It appears the failing tests are for `access-argument` and `access-variable`

* The performance counter report 3 entries and `hpx::agas::find_symbols` query returns the following 8 entries for `access-argument`:
```
/phylanx/access-argument#0#x/0#196
/phylanx/access-argument#0#x/0#262
/phylanx/access-argument#0#x/0#351
/phylanx/access-argument#0#x/0#406
/phylanx/access-argument#0#x/0#464
/phylanx/access-argument#0#x/0#625
/phylanx/access-argument#1#y/0#681
/phylanx/access-argument#2#alpha/0#881
```

* The performance counter report 6 entries and `hpx::agas::find_symbols` query returns the following 14 entries for `access-variable`.
```
"/phylanx/access-variable#0#weights/0#1014"
"/phylanx/access-variable#0#weights/0#628"
"/phylanx/access-variable#0#weights/0#861"
"/phylanx/access-variable#0#weights/0#870"
"/phylanx/access-variable#1#transx/0#757"
"/phylanx/access-variable#2#pred/0#597"
"/phylanx/access-variable#2#pred/0#674"
"/phylanx/access-variable#3#error/0#667"
"/phylanx/access-variable#3#error/0#765"
"/phylanx/access-variable#4#gradient/0#743"
"/phylanx/access-variable#4#gradient/0#889"
"/phylanx/access-variable#5#step/0#537"
"/phylanx/access-variable#5#step/0#931"
"/phylanx/access-variable#5#step/0#937"
```

**Comment by hkaiser** [comment]
> The performance counter /phylanx{locality#0/total}/primitives/access-argument/time/eval reports 3 entries and hpx::agas::find_symbols query returns the following 8 entries for access-argument:

@parsa: yah, that makes sense. 5 of the 8 primitives essentially refer to the same variable. I'm not sure yet how to solve this, though.

File: phylanx/include/util.hpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #define PHYLANX_UTIL_HPP
 
 #include <phylanx/config.hpp>
+#include <phylanx/util/instrumentation/performance_data.hpp>
```
Comment (hkaiser) [inline phylanx/include/util.hpp]: We usually try to match the namespace with the directory structure. I think we can safely put the new header file directly into `phylanx/util` for now to make those match.

File: phylanx/util/instrumentation/performance_data.hpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+{
+    // Retrieve performance counter data for selected primitives
+    //     primitive_instances:
+    //         The primitives for which performance counter data is required
+    //     locality_id:
+    //         The locality the performance counter data is going to be queried from
+    //     counter_name_last_parts:
+    //         A vector containing the last part of the performance counter
+    //         names. e.g. std::vector{ "count/eval", "count/direct_eval" }
+    PHYLANX_EXPORT std::map<std::string, std::vector<std::int64_t>>
```
Comment (hkaiser) [inline phylanx/util/instrumentation/performance_data.hpp]: Could you turn this comment into a proper doxygen formatted one, please?

File: tests/unit/util/performance_data.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    std::vector<std::string> existing_primitive_instances;
+
+    // Retrieve all primitive instances
+    for (auto const& entry :
+        hpx::agas::find_symbols(hpx::launch::sync, "/phylanx/*#*"))
+    {
+        existing_primitive_instances.push_back(entry.first);
+    }
+
+    // TODO: Finish the test
```
Comment (hkaiser) [inline tests/unit/util/performance_data.cpp]: Will you 'finish the test'?

File: phylanx/include/util.hpp (hunk 1 above)
Comment (parsa) [reply phylanx/include/util.hpp]: Fixed.

File: phylanx/util/instrumentation/performance_data.hpp (hunk 2 above)
Comment (parsa) [reply phylanx/util/instrumentation/performance_data.hpp]: Please check again.

File: tests/unit/util/performance_data.cpp (hunk 3 above)
Comment (parsa) [reply tests/unit/util/performance_data.cpp]: I needed to show the `primitive_counter` issue first. The test is finished now.

**Pull request merged**